DigitalEcosystem/publication/feature_engineering/featurize_all.py

```diff
@@ -0,0 +1,346 @@
+# 
+# Here, we'll take the neighbor-finding algorithm of JMol, as implemented in PyMatGen.
+
+# ## Average Bond Length
+# [Documentation Link](https://hackingmaterials.lbl.gov/matminer/matminer.featurizers.site.html#matminer.featurizers.site.bonding.AverageBondLength)
+# 
+# We will also calculate the average bond length in our system.
+
+# In[]:
+
+
+df['bond_length_average'] = df.ox_struct.swifter.apply(average_bond_length).copy()
+
+
+# ## Average Bond Angle
+# [Documentation Link](https://hackingmaterials.lbl.gov/matminer/matminer.featurizers.site.html#matminer.featurizers.site.bonding.AverageBondAngle)
+# 
+# We'll additionally determine the average bond angle between atoms in our system.
+
+# In[]:
+
+
+df['bond_angle_average'] = df.ox_struct.swifter.apply(average_bond_angle).copy()
+
+
+# ## Average Coordination Number
+# 
+# We'll determine how coordinated, on average, the different atoms in our system are. This is accomplished by
+# determining the number of nearest neighbors with the neighbor finder we defined earlier, and averaging over
+# the atoms in the crystal cell.
+
+# In[]:
+
+
+df['average_cn'] = df.ox_struct.swifter.apply(average_cn).copy()
+
+
+# ## AB Perimeter/Area Ratio
+# 
+# Next, we'll try to assess how square our unit cell is in the two directions parallel to the surface. To do this, we'll
+# take the ratio of the perimeter to the area of the cell.
+
+# In[]:
+
+
+df['perimeter_area_ratio'] = df.ox_struct.swifter.apply(ab_perimeter_area_ratio).copy()
+
+
+# Next, we'll extract the formula from the system.
+
+# In[]:
+
+
+df["formula"] = df["atoms_object (unitless)"].swifter.apply(lambda atoms: atoms.get_chemical_formula(empirical=True))
+
+
+# ## Structural Descriptors
+# 
+# Next, we'll calculate some structural descriptors, similar to the ones that had been used by Bhowmik, R. et al in [DOI 10.1016/j.polymer.2021.123558](https://doi.org/10.1016/j.polymer.2021.123558).
+# 
+# Because there are a wide variety of elements in our dataset, to try and keep this from generating too many columns, we'll bin each element into different sections of the periodic table:
+# - Alkaline metals
+# - Alkaline earth metals
+# - Early transition metals (d5 metals and earlier)
+# - Late transition metals (d6 metals and later)
+# - Triels (group 13 / Boron group)
+# - Tetrels (group 14 / Carbon group)
+# - Pnictogens
+# - Chalcogens
+# - Halides
+# - Noble Gases
+# - F-block elements
+# - Post-Uranium elements
+# 
+# We'll be looking for the following counts:
+# - Number of each group (e.g. number of alkaline, number of triel, etc)
+# - Number of each group-group bond (e.g. number of alkaline-early_transition, number of chalcogen-chalcogen, etc)
+# - Number of each grou-grou-group angle (e.g. number of halide-late_transition-halide, number of tetrel-tetrel-tetrel, etc)
+# 
+# Adding dihedrals as well would have required a more complex graph traversal algorithm, so in the interest of time we forego this.
+
+# In[]:
+
+
+classes = {
```

Comment:
I put it here since it never gets re-used elsewhere - justification being that it made sense at the time to define a one-time-use variable near it gets used. I can toss it somewhere into utils though, probably wherever we put element definitions